wallet-new/src/Cardano/Wallet/Kernel/DB/Sqlite.hs

```diff
@@ -436,75 +473,172 @@ instance Ord OrdByCreationDate where
+        Left e -> throwIO $ Kernel.StorageFailure (toException e)
+        Right Nothing -> return ([], Just 0)
+        Right (Just (meta, inputs, outputs)) ->  do
+            eiCount <- limitExecutionTimeTo (25 :: Second) (\ _ -> ()) $ mapLeft $ Sqlite.runDBAction $ runBeamSqlite conn $
+                case mbAddress of
+                    Nothing   -> SQL.runSelectReturningOne $ SQL.select metaQueryC
+                    Just addr -> SQL.runSelectReturningOne $ SQL.select $ metaQueryWithAddrC addr
+            let mapWithInputs  = transform $ map (\inp -> (unlockTxId $ _getTxInput inp, inp)) inputs
+            let mapWithOutputs = transform $ map (\out -> (unlockTxId $ _getTxOutput out, out)) outputs
+            let txMeta = toValidKernelTxMeta mapWithInputs mapWithOutputs $ Data.List.NonEmpty.toList meta
+                count = case eiCount of
+                    Left _  -> Nothing
+                    Right c -> c
+            return (txMeta, count)
+
     where
-        getTx selector = _txCoinDistributionTxId . selector
-
-        metaQuery = case mbSorting of
-            Nothing  -> SQL.all_ (_mDbMeta metaDB)
-            Just (Sorting SortByCreationAt dir) ->
-                SQL.orderBy_ (toBeamSortDirection dir . _txMetaTableCreatedAt) $ SQL.all_ (_mDbMeta metaDB)
-            Just (Sorting SortByAmount     dir) ->
-                SQL.orderBy_ (toBeamSortDirection dir . _txMetaTableAmount) $ SQL.all_ (_mDbMeta metaDB)
-
-        -- The following two queries are disjointed and both fetches, respectively,
-        -- a list of tuples of type @(TxMeta, TxInput)@ and @(TxMeta, TxOutput)@.
-        -- The rationale behind doing two separate queries is that there is no elegant
-        -- way to express a 3-table join without having 'Maybe's cropping up in the
-        -- Haskell result sets, and furthermore we would have to deal with duplicates.
-        -- Doing two separate queries is yes more I/O taxing, but spares us from doing
-        -- duplicate filtering on the Haskell side.
-        paginatedInputs = SQL.select $ do
-            meta   <- SQL.limit_ limit (SQL.offset_ offset metaQuery)
-            inputs  <- SQL.oneToMany_ (_mDbInputs metaDB)  (getTx _getTxInput) meta
-            pure (meta, inputs)
-
-        paginatedOutputs = SQL.select $ do
-            meta   <- SQL.limit_ limit (SQL.offset_ offset metaQuery)
-            outputs <- SQL.oneToMany_ (_mDbOutputs metaDB) (getTx _getTxOutput) meta
-            pure (meta, outputs)
-
-        -- | Groups the inputs or the outputs under the same 'TxMeta'.
-        transform :: NonEmpty (TxMeta, a) -> M.Map OrdByCreationDate (NonEmpty a)
+        mapLeft :: IO (Either a b) -> IO (Either () b)
+        mapLeft m  = do
+            x <- m
+            case x of
+                Left _  -> return $ Left ()
+                Right r -> return $ Right r
+
+        filters meta = do
```

Comment:
I'd personally appreciate type signatures for these local definitions but perhaps beam makes that too awkward?

Reply:
Awkward indeed. filters has this signature : 
‘TxMetaT
                                          (SQL.QExpr
                                             SqliteExpressionSyntax
                                             (SQL.QNested
                                                (SQL.QNested Database.Beam.Query.QueryInaccessible)))
                                        -> SQL.Q
                                             SqliteSelectSyntax
                                             MetaDB
                                             (SQL.QNested
                                                (SQL.QNested Database.Beam.Query.QueryInaccessible))
                                             a1’

Reply:
Ok, point taken :joy: